# Data 충분해요?

* writer : rayleigh kim

**답 들이 적혀있습니다. 강의자는 답 들을 지우고 배포해주세요.**

**First things first!**<br>
지금 실습할 데이터의 상황에 대한 **아주 구체적인 시나리오**가 안잡혀 있으면, 실습의 의미가 희미해집니다.

**시각화를 복습할 시간입니다.**<br>
1. 데이터가 충분한지 확인합시다.
2. 부족하다면, 과연 아무 것도 할 수 없을까요? EDA토론 시간을 떠올려봅시다.
3. 그럼에도, 뭐가 어떻게 부족한지 알아야만 **적절한 실험계획을 세웁니다.**


## 데이터 로딩

In [1]:
import pandas as pd

In [2]:
temp2 = pd.read_csv('ade_temp2.csv')

In [3]:
temp2.shape

(31, 23)

In [4]:
temp2.head()

Date Location  Lemon  Orange  Temperature  Leaflets  Price  dayname  \
0  2016-07-01     Park     97      67           70      90.0   0.25        4   
1  2016-07-02     Park     98      67           72      90.0   0.25        5   
2  2016-07-03     Park    110      77           71     104.0   0.25        6   
3  2016-07-04    Beach    134      99           76      98.0   0.25        0   
4  2016-07-05    Beach    159     118           78     135.0   0.25        1   

   lemon_profit  orange_profit     ...       orange_prepared lemon_total_cost  \
0         24.25          16.75     ...                   150             10.0   
1         24.50          16.75     ...                   150             10.0   
2         27.50          19.25     ...                   150             10.0   
3         33.50          24.75     ...                   150             10.0   
4         39.75          29.50     ...                   150             10.0   

   orange_total_cost  net_profit  x_orange  x_lemon  loss_orange  loss_lemon  \
0                7.5       18.50        83      103         4.15        5.15   
1                7.5       18.75        83      102         4.15        5.10   
2                7.5       24.25        73       90         3.65        4.50   
3                7.5       35.75        51       66         2.55        3.30   
4                7.5       46.75        32       41         1.60        2.05   

   loss_total  new_factor1  
0        9.30            H  
1        9.25            H  
2        8.15            H  
3        5.85            O  
4        3.65            O  

[5 rows x 23 columns]

In [5]:
temp2.tail()

Date Location  Lemon  Orange  Temperature  Leaflets  Price  dayname  \
26  2016-07-27     Park    104      68           80      99.0   0.35        2   
27  2016-07-28     Park     96      63           82      90.0   0.35        3   
28  2016-07-29     Park    100      66           81      95.0   0.35        4   
29  2016-07-30    Beach     88      57           82      81.0   0.35        5   
30  2016-07-31    Beach     76      47           82      68.0   0.35        6   

    lemon_profit  orange_profit     ...       orange_prepared  \
26          36.4          23.80     ...                   150   
27          33.6          22.05     ...                   150   
28          35.0          23.10     ...                   150   
29          30.8          19.95     ...                   150   
30          26.6          16.45     ...                   150   

   lemon_total_cost  orange_total_cost  net_profit  x_orange  x_lemon  \
26             10.0                7.5       37.70        82       96   
27             10.0                7.5       33.15        87      104   
28             10.0                7.5       35.60        84      100   
29             10.0                7.5       28.25        93      112   
30             10.0                7.5       20.55       103      124   

    loss_orange  loss_lemon  loss_total  new_factor1  
26         4.10         4.8        8.90            H  
27         4.35         5.2        9.55            H  
28         4.20         5.0        9.20            H  
29         4.65         5.6       10.25            O  
30         5.15         6.2       11.35            O  

[5 rows x 23 columns]

In [6]:
temp2.describe()

Lemon      Orange  Temperature    Leaflets      Price    dayname  \
count   31.000000   31.000000    31.000000   31.000000  31.000000  31.000000   
mean   116.580645   80.354839    78.870968  109.129032   0.358065   3.225806   
std     26.136532   22.130745     4.096681   19.822112   0.113355   2.060901   
min     71.000000   42.000000    70.000000   68.000000   0.250000   0.000000   
25%     98.000000   66.500000    77.000000   92.500000   0.250000   1.500000   
50%    115.000000   77.000000    80.000000  108.000000   0.350000   3.000000   
75%    132.500000   95.000000    82.000000  124.000000   0.500000   5.000000   
max    176.000000  129.000000    84.000000  158.000000   0.500000   6.000000   

       lemon_profit  orange_profit  total_profit  lemon_prepared  \
count     31.000000      31.000000     31.000000            31.0   
mean      40.967742      28.000000     68.967742           200.0   
std       14.343724      10.346505     24.614506             0.0   
min       24.250000      16.450000     41.000000           200.0   
25%       30.775000      21.125000     52.000000           200.0   
50%       35.500000      24.500000     59.500000           200.0   
75%       52.500000      34.250000     86.750000           200.0   
max       78.000000      56.500000    134.500000           200.0   

       orange_prepared  lemon_total_cost  orange_total_cost  net_profit  \
count             31.0              31.0               31.0   31.000000   
mean             150.0              10.0                7.5   46.467742   
std                0.0               0.0                0.0   24.614506   
min              150.0              10.0                7.5   18.500000   
25%              150.0              10.0                7.5   29.500000   
50%              150.0              10.0                7.5   37.000000   
75%              150.0              10.0                7.5   64.250000   
max              150.0              10.0                7.5  112.000000   

         x_orange     x_lemon  loss_orange  loss_lemon  loss_total  
count   31.000000   31.000000    31.000000   31.000000   31.000000  
mean    69.645161   83.419355     3.482258    4.170968    7.653226  
std     22.130745   26.136532     1.106537    1.306827    2.411394  
min     21.000000   24.000000     1.050000    1.200000    2.250000  
25%     55.000000   67.500000     2.750000    3.375000    6.150000  
50%     73.000000   85.000000     3.650000    4.250000    7.950000  
75%     83.500000  102.000000     4.175000    5.100000    9.275000  
max    108.000000  129.000000     5.400000    6.450000   11.850000

In [7]:
temp2.dtypes

Date                  object
Location              object
Lemon                  int64
Orange                 int64
Temperature            int64
Leaflets             float64
Price                float64
dayname                int64
lemon_profit         float64
orange_profit        float64
total_profit         float64
temper_group          object
lemon_prepared         int64
orange_prepared        int64
lemon_total_cost     float64
orange_total_cost    float64
net_profit           float64
x_orange               int64
x_lemon                int64
loss_orange          float64
loss_lemon           float64
loss_total           float64
new_factor1           object
dtype: object

In [8]:
# 코드 실행 후, 날짜 반드시 확인.  최악의 경우 기나긴 코딩/수작업이 필요하다.
temp2['Date'] = pd.to_datetime(temp2['Date'])

In [9]:
temp2.head()

Date Location  Lemon  Orange  Temperature  Leaflets  Price  dayname  \
0 2016-07-01     Park     97      67           70      90.0   0.25        4   
1 2016-07-02     Park     98      67           72      90.0   0.25        5   
2 2016-07-03     Park    110      77           71     104.0   0.25        6   
3 2016-07-04    Beach    134      99           76      98.0   0.25        0   
4 2016-07-05    Beach    159     118           78     135.0   0.25        1   

   lemon_profit  orange_profit     ...       orange_prepared lemon_total_cost  \
0         24.25          16.75     ...                   150             10.0   
1         24.50          16.75     ...                   150             10.0   
2         27.50          19.25     ...                   150             10.0   
3         33.50          24.75     ...                   150             10.0   
4         39.75          29.50     ...                   150             10.0   

   orange_total_cost  net_profit  x_orange  x_lemon  loss_orange  loss_lemon  \
0                7.5       18.50        83      103         4.15        5.15   
1                7.5       18.75        83      102         4.15        5.10   
2                7.5       24.25        73       90         3.65        4.50   
3                7.5       35.75        51       66         2.55        3.30   
4                7.5       46.75        32       41         1.60        2.05   

   loss_total  new_factor1  
0        9.30            H  
1        9.25            H  
2        8.15            H  
3        5.85            O  
4        3.65            O  

[5 rows x 23 columns]

## 데이터가 충분할까요?

물론 row가 31개 뿐이라는 것은 잘 알고 있습니다.
터무니 없이 부족해요, 그래도 통계학자란 12개만 넘으면 어떻게든 해보는 사람들입니다.

하지만...

**데이터가 부족하냐 / 충분하냐 는 단순히 rows 숫자를 그대로 보면 안됩니다.**

**Lemon, Orange에 대해서 각각 혹은 하나를 택하여 실습을 진행합니다.**

### 실습 C.1

1. 전체 row 중 Park 가 몇 번인지, Beach가 몇번인지 구하시오.
2. 효과적인 시각화 방법을 의논하고, 시각화 하시오.
3. 무엇이 문제가 될 수 있을지 토론하시오.

### 실습 C.2

1. Price 0.25로 몇 번, 0.35로 몇 번, 0.5 로 몇 번 팔았는지 구하시오.
2. 효과적인 시각화 방법을 의논하고, 시각화 하시오.
3. 무엇이 문제가 될 수 있을지 토론하시오.

### 실습 C.3

1. Price와 Location을 이용하여 새로운 조합을 만듭니다. 아래의 코드를 사용하면 편합니다.
```
temp2['PL'] = temp2['Location'] +'_'+ temp2['Price'].map(str)
```
2. 새로운 변수(factor) PL안의 조합별로 각각 row가 몇 번 있나요?
3. 무엇이 문제가 될 수 있을까요?

### 실습 C.4

1. 묶은 세로 막대형 차트를 사용할 겁니다.
2. Price 별로, Lemon에이드를 '평균' 몇 개 팔았는지 시각화 하세요.
3. 범례로 Location을 두어서, 장소에 따라서 다른 막대가 표시되도록 하세요.

Q1. Location이 범례로 있기 전에는(2번 차트), 어떤 Price로 팔아야 가장 많이 팔았나요?

Q2. Location이 범례로 있을 때는(3번 차트), 장소별로 어떤 Price가 가장 많이 팔았나요?

Q3. 이는, 여러분에게 '도움'을 주는 적절한 시각화 인가요?
1. 어떤 측면에서는 도움이 되었을까요
2. 어떤 측면에서는 한계가 있을까요

### 실습 C.5

1. 변수 PL 안의 조합별로 레몬에이드 판매량의 평균을 구하세요.
2. 어떤 시각화 방법이 효율적일까요? 시각화하세요.
3. 실습 C.4 와 어떤 차이점이 있을까요?
    * 어떠한 상황에서 C.4의 방법이 효율적일까요?
    * 어떠한 상황에서 C.5의 방법이 효율적일까요?

### 실습 C.6

1. 레몬에이드 판매량, 오렌지에이드 판매량, 온도, 리플렛에 대하여 각각 진행합니다.
2. 각각을 살펴보기에 가장 좋은 시각화 방법을 의논하고, 시각화 하세요.
    * 레몬에이드만을 두고 예를 들자면, 레몬에이드변수만 가지고 시각화 할 수도 있어요.
    * 시계열 데이터입니다. 시간에 따라 레몬에이드가 어떻게 변화되는지 관찰할 수도 있습니다.
    
3. 레몬에이드 판매량, 오렌지에이드 판매량, 온도, 리플렛의 [평균, 중앙값, 최대값, 최소값, 표준편차]를 구하세요.

4. 온도, 리플렛 의 변동범위 안에서 데이터가 유니폼하게 분포되어있나요? histogram이 flat
    * 고르게 분포되어 있다라는 것은 데이터가 옹기종기 모여있다는 뜻입니다.
    * 유니폼하게 분포되어 있다라는 것은, 여기저기 잘 흩어져 있다는 말입니다.
    * Q1. 온도와 리플렛이 한군데에 뭉쳐져 있으면 무슨 장/단점이 있을까요?
    * Q2. 온도와 리플렛이 여기저기 흩어져 있으면 무슨 장/단점이 있을까요?
    * Q3. 시각화 없이 단순하게 알아볼만한 방법이 뭐가 있을까요? 표준편차, 첨도 보면 된다.
    
5. 판매량에 대하여 4번 문제를 고민합시다. 데이터의 변동이 큰가요 작은가요?
    * 크다고 판단했다면, 왜 인가요?
    * 작다고 판단했다면, 왜 인가요?
    
    